cedar-lean/Cedar/Thm/Partial/Evaluation/Record.lean

```diff
@@ -148,4 +149,158 @@ theorem on_concrete_eqv_concrete_eval {attrs : List (Attr × Spec.Expr)} {reques
+    have ⟨ih_hd, ih_tl⟩ := ih ; clear ih
+    have (khd, xhd) := hd ; clear hd
+    simp only at *
+    cases h₃ : Partial.evaluate xhd req entities
+    <;> simp only [h₃, Except.bind_err, Except.bind_ok] at h₁
+    case ok hd_pval =>
+      unfold is_all_concrete at h₂
+      replace ⟨vs, h₂⟩ := h₂
+      replace ⟨h₂, h₂'⟩ := And.intro (List.mapM_some_implies_all_some h₂) (List.mapM_some_implies_all_from_some h₂)
+      cases h₅ : tl.mapM (λ kv => do let v ← Partial.evaluate kv.snd req entities ; .ok (kv.fst, v))
+      <;> simp only [h₅, Except.bind_ok, Except.ok.injEq, Except.bind_err] at h₁
+      case ok tl_pvals =>
+        subst h₁
+        cases h₄ : Partial.evaluate (xhd.subst subsmap) req' (entities.subst subsmap)
+        <;> simp only [Except.bind_err, Except.bind_ok]
+        case error e =>
+          replace ⟨v, _, h₂⟩ := h₂ hd_pval (by simp)
+          cases hd_pval <;> simp only [Option.some.injEq] at h₂
+          case value v' =>
+            subst v'
+            unfold SubstPreservesEvaluationToConcrete at ih_hd
+            simp only [ih_hd h_req v h₃] at h₄
+        case ok hd'_pval =>
+          have ih₂ := mapM_subst_preserves_evaluation_to_values ih_tl h_req tl_pvals h₅ (by
+            unfold is_all_concrete
+            apply List.all_some_implies_mapM_some
+            intro tl_pval h₆
+            replace ⟨v, _, h₂⟩ := h₂ tl_pval (by simp [h₆])
+            exists v
+          )
+          simp only [ih₂, Except.bind_ok, Except.ok.injEq, List.cons.injEq, Prod.mk.injEq,
+            true_and, and_true]
+          cases hd_pval <;> simp only [List.map_cons, List.mem_cons, List.mem_map, forall_eq_or_imp,
+            and_false, false_and, exists_const, forall_exists_index, and_imp, forall_apply_eq_imp_iff₂,
+            Option.some.injEq] at h₂
+          case value hd_val =>
+            unfold SubstPreservesEvaluationToConcrete at ih_hd
+            simp only [ih_hd h_req hd_val h₃, Except.ok.injEq] at h₄
+            exact h₄.symm
+
+private theorem lemma {pvals : List (Attr × Partial.Value)} {pairs : List (Attr × Spec.Value)}:
+  pvals.mapM (λ kv => match kv.snd with
+      | .value v => some (kv.fst, v)
+      | .residual _ => none)
+    = some pairs →
+  pvals.mapM (λ kv => match kv.snd with
+      | .value v => some v
+      | .residual _ => none)
+    = some (pairs.map Prod.snd)
+:= by
+  intro h₁
+  cases pvals <;> simp only [List.mapM_nil, List.mapM_cons, Option.pure_def,
+    Option.bind_eq_bind, Option.bind_eq_some, Option.some.injEq] at *
+  case nil => subst h₁ ; simp only [List.map_nil]
+  case cons hd tl =>
+    have (khd, vhd) := hd ; clear hd
+    simp only at *
+    replace ⟨(khd', vhd'), h₁, h₂⟩ := h₁
+    cases vhd <;> simp only [Option.some.injEq, Prod.mk.injEq, exists_eq_left'] at *
+    replace ⟨h₁, h₁'⟩ := h₁ ; subst khd' vhd' ; rename_i vhd
+    replace ⟨tl', h₂, h₃⟩ := h₂
+    subst h₃
+    exists (tl'.map Prod.snd)
+    simp only [List.map_cons, and_true]
+    exact lemma h₂
```

Comment:
Nit: I was thrown off by the name `lemma` for this theorem, and read this line as a special form of the `exact` tactic (as in `exact lemma` vs just `exact`).

Reply:
Yeah, I should come up with a better name for this lemma

Reply:
Renamed. Let me know if you have a better suggestion